sql/org_mozilla_vrbrowser_derived/clients_daily_v1/query.sql

```diff
@@ -0,0 +1,169 @@
+CREATE TEMP FUNCTION udf_geo_struct(
+  country STRING,
+  city STRING,
+  geo_subdivision1 STRING,
+  geo_subdivision2 STRING
+) AS ( IF(
+    country IS NULL
+    OR country = '??',
+    NULL,
+    STRUCT(
+      country,
+      NULLIF(city, '??') AS city,
+      NULLIF(geo_subdivision1, '??') AS geo_subdivision1,
+      NULLIF(geo_subdivision2, '??') AS geo_subdivision2
+    )
+  )
+);
+
+CREATE TEMP FUNCTION
+  udf_glean_timespan_seconds(timespan STRUCT<time_unit STRING, value INT64>)
+  RETURNS INT64 AS (
+    CAST(
+      FLOOR(
+        CASE timespan.time_unit
+          WHEN 'nanosecond' THEN timespan.value / 1000 / 1000 / 1000
+          WHEN 'microsecond' THEN timespan.value / 1000 / 1000
+          WHEN 'millisecond' THEN timespan.value / 1000
+          WHEN 'second' THEN timespan.value
+          WHEN 'minute' THEN timespan.value * 60
+          WHEN 'hour' THEN timespan.value * 60 * 60
+          WHEN 'day' THEN timespan.value * 60 * 60 * 24
+        END )
+      AS INT64));
+
+CREATE TEMP FUNCTION udf_json_mode_last(list ANY TYPE) AS (
+  (
+    SELECT
+      ANY_VALUE(_value)
+    FROM
+      UNNEST(list) AS _value
+    WITH OFFSET AS _offset
+    GROUP BY
+      TO_JSON_STRING(_value)
+    ORDER BY
+      COUNT(_value) DESC,
+      MAX(_offset) DESC
+    LIMIT
+      1
+  )
+);
+
+CREATE TEMP FUNCTION
+  udf_mode_last(list ANY TYPE) AS ((
+    SELECT
+      _value
+    FROM
+      UNNEST(list) AS _value
+    WITH
+    OFFSET
+      AS
+    _offset
+    GROUP BY
+      _value
+    ORDER BY
+      COUNT(_value) DESC,
+      MAX(_offset) DESC
+    LIMIT
+      1 ));
+
+WITH unioned AS (
+  SELECT
+    submission_timestamp,
+    document_id,
+    client_info,
+    sample_id,
+    metadata,
+    normalized_channel,
+    metrics AS baseline_metrics,
+    NULL AS metrics
+  FROM
+    org_mozilla_vrbrowser_stable.baseline_v1
+  UNION ALL
+  SELECT
+    submission_timestamp,
+    document_id,
+    client_info,
+    sample_id,
+    metadata,
+    normalized_channel,
+    NULL AS baseline_metrics,
+    metrics
+  FROM
+    org_mozilla_vrbrowser_stable.metrics_v1
```

Comment:
I looked into using `LEFT JOIN`, the best I could come up with but also doesn't quite work was:
```diff
   FROM
     org_mozilla_vrbrowser_stable.baseline_v1
   UNION ALL
+  -- Consider users only as active if they sent a baseline ping.
   SELECT
-    submission_timestamp,
-    document_id,
-    client_info,
-    sample_id,
-    metadata,
-    normalized_channel,
+    metrics.submission_timestamp,
+    metrics.document_id,
+    metrics.client_info,
+    metrics.sample_id,
+    metrics.metadata,
+    metrics.normalized_channel,
     NULL AS baseline_metrics,
-    metrics
+    metrics.metrics
   FROM
-    org_mozilla_vrbrowser_stable.metrics_v1
+    org_mozilla_vrbrowser_stable.baseline_v1 AS baseline
+  LEFT JOIN
+    org_mozilla_vrbrowser_stable.metrics_v1 AS metrics
+  ON
+    baseline.client_info.client_id = metrics.client_info.client_id
+  WHERE
+    baseline.client_info.client_id IS NOT NULL
 ),
   --
 base AS (
```

Reply:
Good point. I think we'll need to do separate aggregations on the baseline and metrics pings and then do the join between the two CTEs that each have only one row per client.

This also touches on a broader issue that we'd like to be able to support new glean applications automatically for GUD, etc. The baseline ping is standard for all apps and is intended as the source of being able to calculate retention, etc. If we need to aggregate baseline and metrics separately anyway to do the join here, maybe the best way forward here is to define a completely generic `baseline_daily` query and then individual apps can define custom `metrics_daily` queries that can be joined in as needed.

So, if it seems like doing separate aggregations and then joining is a straight-forward solution here, let's do that for now. Otherwise, we can table this for the moment and discuss the more general way forward for glean aggregations during Berlin and include @fbertsch in the discussion.

Reply:
I we do separate aggregations would we simply use dimensions related to `client_info` from `baseline_daily`? That behaviour (and potentially event results?) would be different from using `udf_mode_last` on both baseline and metrics pings.

Reply:
We would used `udf_mode_last` over baseline pings for the fields that exist in baseline pings, and we would use `udf_mode_last` over metrics pings for any metrics-ping-specific fields.

Yes, that could lead to us choosing a different value for a dimension compared to using `udf_mode_last` over baseline and metrics pings. But it seems conceptually reasonable to me to accept the value we see from the baseline ping. Baseline pings become the main source of truth, and metrics pings are considered only for enriching the data with dimensions that exist as metrics.